File: 백준/Silver/2579. 계단 오르기/계단 오르기.java
import java.io.*;
import java.util.*;

public class Main
{
    static int[] dp;
    static int[] a;
	public static void main(String[] args) throws IOException {
	    BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
	    int n=Integer.parseInt(br.readLine());
	    
	    dp=new int[n+1];
	    a=new int[n+1];

	    
	    for(int i=1;i<=n;i++){
	        a[i]=Integer.parseInt(br.readLine());
	    }
	    
	    dp[1]=a[1];
	    
	    if(n<3) {
	        int sum=0;
	        for(int i:a) sum+=i;
	        System.out.println(sum);
	        return;
	    }
	    
	    dp[2]=dp[1]+a[2];
	    
	    for(int i=3;i<=n;i++){
	        dp[i]=Math.max(dp[i-2],dp[i-3]+a[i-1])+a[i];
	    }
	    
		System.out.println(dp[n]);
	}
}

File: 백준/Silver/2579. 계단 오르기/계단 오르기.py
import sys
input=sys.stdin.readline

N=int(input())
a=[int(input()) for _ in range(N)]
dp=[0 for _ in range(N)]

if N<=2: 
    print(sum(a))
    exit()
    
else:
    dp[0]=a[0]
    dp[1]=a[0]+a[1]
    dp[2]=max(a[0],a[1])+a[2]

    for i in range(3,N):
        dp[i]=max(dp[i-2]+a[i],dp[i-3]+a[i-1]+a[i])
    
    print(dp[N-1])